identity-iota/src/did/doc/iota_document.rs

```diff
@@ -1601,6 +1650,90 @@ mod tests {
     assert_eq!(document.methods().count(), 1);
   }
 
+  #[test]
+  fn test_attach_verification_relationships() {
+    let keypair: KeyPair = generate_testkey();
+    let mut document: IotaDocument = IotaDocument::new(&keypair).unwrap();
+
+    let fragment = "#attach-test";
+    let method = iota_verification_method(document.did().as_ref(), fragment);
+    document.insert_method(method, MethodScope::VerificationMethod);
+
+    assert!(document
+      .attach_method_relationship(
+        document.did().to_url().join(fragment).unwrap(),
+        MethodRelationship::CapabilityDelegation,
+      )
+      .is_ok());
+
+    assert_eq!(document.as_document().verification_relationships().count(), 2);
+
+    // Adding it a second time does nothing.
+    assert!(document
+      .attach_method_relationship(
+        document.did().to_url().join(fragment).unwrap(),
+        MethodRelationship::CapabilityDelegation,
+      )
+      .is_ok());
+
+    // len is still 2.
+    assert_eq!(document.as_document().verification_relationships().count(), 2);
+
+    // Attempting to attach a relationship to a non-existing method fails.
+    assert!(document
+      .attach_method_relationship(
+        document.did().to_url().join("#doesNotExist").unwrap(),
+        MethodRelationship::CapabilityDelegation,
+      )
+      .is_err());
+  }
+
+  #[test]
+  fn test_detach_verification_relationships() {
+    let keypair: KeyPair = generate_testkey();
+    let mut document: IotaDocument = IotaDocument::new(&keypair).unwrap();
+
+    let fragment = "#detach-test";
+    let method = iota_verification_method(document.did().as_ref(), fragment);
+    document.insert_method(method, MethodScope::VerificationMethod);
+
+    assert!(document
+      .attach_method_relationship(
+        document.did().to_url().join(fragment).unwrap(),
+        MethodRelationship::AssertionMethod,
+      )
+      .is_ok());
+
+    assert!(document
+      .detach_method_relationship(
+        document.did().to_url().join(fragment).unwrap(),
+        MethodRelationship::AssertionMethod,
+      )
+      .is_ok());
+
+    // len is 1; the relationship was removed.
+    assert_eq!(document.as_document().verification_relationships().count(), 1);
+
+    // Removing it a second time does nothing
+    assert!(document
+      .detach_method_relationship(
+        document.did().to_url().join(fragment).unwrap(),
+        MethodRelationship::AssertionMethod,
+      )
+      .is_ok());
```

Comment:
I think we had some agreement on moving the existence and duplication checks for `insert_method` and `remove_method` to the lower-level APIs. From that perspective, I suppose the relationship methods should do the same? If it's fine for now, then I would address this together with moving the method checks in another PR, as this is a slightly larger change, affecting error types, the `IotaDocument` as well as the account updates.

In the meantime, I'll change `detach_method_relationship` to return `Result<bool>` for at least some consistency.

Reply:
> In the meantime, I'll change detach_method_relationship to return Result<bool> for at least some consistency.
>
> Edit: Can't change the return type, because OrderedSet::remove doesn't return bool.

I think we can check for existence in `OrderedSet::remove` first, much like `OrderedSet::append`:

```rust
  /// Adds a new value to the end of the `OrderedSet`; returns `true` if the
  /// value was successfully added.
  pub fn append(&mut self, item: T) -> bool
  where
    T: KeyComparable,
  {
    if self.contains(&item) {
      false
    } else {
      self.0.push(item);
      true
    }
  }
```